### DISCUSSION TITLE
Creating and Running commands with fastapi (manager commands)

### DISCUSSION THREAD

**Discussion by aherle** [fastapi/fastapi#7785] [Questions] [answered]
#### Is it possible to create a command and running it using FastAPI ?

i..e. Is FastAPI allow execute manager commands like what we have in flask ?

Referece: https://flask-script.readthedocs.io/en/latest/#creating-and-running-commands

**Comment by nkhitrov** [comment]
Hi, @ankithherle !

As I know, FastAPI has not manager, but it has [`startup`](https://fastapi.tiangolo.com/advanced/testing-events/) and `shutdown` events.

**Comment by nsidnev** [comment, marked as answer]
@ankithherle 
Hi! No, `FastAPI` does not provide such functionality out of the box. Just like it does not have built-in support for various ORM or other tools that can bind you to something(except for `pydantic` for validation).

So that you can choose something that you have used before or would like to try. For example, you can take a look at [typer](https://github.com/tiangolo/typer), a click-based library for CLI, from the creator of `FastAPI`. Well, or any other tool, for example, [click](https://github.com/pallets/click/) itself, [cleo](https://github.com/sdispater/cleo), or many other libraries.

**Comment by tiangolo** [comment]
Thanks for the help here @Slyfoxy and @nsidnev ! :cake: :bow: 

@ankithherle does that solve your question? May we close this issue?

**Comment by aherle** [comment]
Yes, my issue is solved, Thanks.

**Comment by khushalt** [comment]
How to register typer cli commands in FastAPI app
@tiangolo

**Reply by simonsax** [reply]
I have the same question and would love to know how to integrate typer with fastapi.

In flask I could register a command easily with the flask app and run the command for example with `flask command setup_db`

```
@app.cli.command(name='setup_db')
def reset_and_setup_db():
    db.drop_all()
    db.create_all()
```

**Reply by Garwend** [reply]
This solution works but I'm not sure if it's the best way to implement it.

**database.py**
```
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

from config import settings

SQLALCHEMY_DATABASE_URL = f"postgresql://{settings.db_user}:{settings.db_password}@{settings.db_host}:{settings.db_port}/{settings.db_name}"

engine = create_engine(SQLALCHEMY_DATABASE_URL)

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

```

**manage.py**
```
import typer
from database import get_db
from sqlalchemy.orm import Session
from models import User
from routers.users import get_password_hash

app = typer.Typer()


@app.command()
def hello(txt: str):
    print(f'HELLO {txt}')


@app.command()
def create_admin(name: str, password: str):
    db: Session = next(get_db())

    user = User()
    user.username = name
    user.password = get_password_hash(password)
    user.is_admin = True

    db.add(user)
    db.commit()
    db.refresh(user)

    print(f'Created admin user {name}')


if __name__ == "__main__":
    app()
```

`python manage.py create-admin admin password123`
